### PR TITLE
Fix race conditions in localized SSR

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes a serious bug in localized server-side rendering discovered by @sgomes when testing the `/log-in` routes. It's caused by the fact that there is one global singleton instance of the `i18n` object on the server and all requests from all users share it.

1. Locally, run `yarn build-languages` so that your `public/languages/` folder is populated with language files like `id-v.1.1.json`.
2. Then run local Calypso with `yarn start`.
3. Disable JavaScript in your browser devtools so that you are sure that you always see only the server-rendered markup and that the client-side JavaScript doesn't run.
4. Open the localized `/id/themes` page. You should see Calypso localized in Indonesian.
5. Now open the `/log-in` page. As it doesn't have any locale suffix (like `/log-in/id`) it should be in English. But it's in Indonesian, the language of the previously rendered Themes page!

Screenshot 2022-01-10 at 12 24 15

The reason is that the `ssrSetupLocale` page.js handler, used on the `/:lang/themes` page, will set the global `i18n` instance to Indonesian locale. Then, when the `/log-in` page is rendered, it doesn't modify the `i18n` instance and uses it, assuming it's using the default `en` locale`. But it doesn't. It uses the somewhat random locale that the last `/themes` request happened to request.

The fix has several parts:
- first, I had to clean up `i18n-calypso` TS types so that they declare the `I18N` class. TypeScript files wouldn't compile otherwise.
- then I added support for React context to `i18n-calypso`. The `localize` HOC and the `useTranslate` hook no longer use the global `i18n` instance directly, but instead use the one provided by `I18NContext`. The default global `i18n` instance is the default value of that context, so that unless you mount a provider, the library behaves exactly the same way as it did before.
- another thing I had to modify is the `/theme-filters` REST request issued by the `/themes` SSR. There is a `wpcom` handler that adds the `locale` query parameter to it, reading the locale value from the global `i18n` instance. That's going to always be `en` now. But we need to pass the real locale, because the filters response contains localized labels. I achieved this by adding a custom `locale` parameter to the `requestThemeFilters` action that overrides the `locale` param.
- finally I modified the `ssrSetupLocale` handler and the rendering code to create a local `i18n` instance and pass it in context.

**How to test:**
Verify that the bug described above is indeed fixed: the `/log-in` page is always English, no matter what request you did before.

Also verify that the `/id/themes` page continues to be localized:

Screenshot 2022-01-10 at 11 28 36

The labels on the toolbar (Majalah, Bisnis) come from the `/theme-filters?locale=id` server response, not from local `translate` calls, so verify that both sources of localization work correctly.

### CODE REVIEW

**Comment by jsnajdr** [comment]
One open issue about this patch: if a React component uses the global `import i18n from 'i18n-calypso'` object and calls `i18n.translate()` on it, these strings won't be localized because the server-side global `i18n` is always `en`. Only the `translate` calls from the `localize`/`useTranslate` React API are translated.

**Comment by sgomes** [comment]
> One open issue about this patch: if a React component uses the global `import i18n from 'i18n-calypso'` object and calls `i18n.translate()` on it, these strings won't be localized because the server-side global `i18n` is always `en`. Only the `translate` calls from the `localize`/`useTranslate` React API are translated.

@jsnajdr : This sounds like those calls will need to be migrated away from that pattern, which depending on how many there are could take a while. And it could cause significant disruption to translations if this patch gets merged before that.

In the meantime, is there any way we could preserve the current `i18n` global-setting behaviour on the client side only? That would reduce the surface area for the change while fixing the server-side issue immediately, and would give us some breathing space to slowly migrate the code that relies on the global `i18n`.

**Comment by jsnajdr** [comment]
> In the meantime, is there any way we could preserve the current `i18n` global-setting behaviour on the client side only?

That's what already happens in this PR: it sets a custom instance only for SSR, in the `ssrSetupLocale` function. The browser code doesn't set the `i18n` anywhere, and the `CalypsoI18Provider` always uses the global instance.

So the only thing that is affected are the server-side localized `/themes` and `/theme` pages, where direct `i18n.translate()` calls will return the original string instead of the translated ones. That's a quite small scope where we can hopefully migrate away from `i18n.translate` quickly.

**Comment by sgomes** [comment]
> That's what already happens in this PR [...]

I see, thank you @jsnajdr, that sounds perfect! 👍 Sorry I misunderstood what's happening here, as I haven't yet been able to fully grok internationalisation in Calypso as a whole.

**Review by sgomes — APPROVED** [review]
Thank you for this extensive PR that fixes such a deep-seated bug!

I haven't found anything problematic with these changes, but my understanding of the whole i18n machinery in Calypso is somewhat limited.

I'm approving this to unblock you in case you're waiting for an approval, @jsnajdr, but I'd feel more comfortable if someone else with more Calypso i18n experience were to review this as well 🙂

**Comment by jsnajdr** [comment]
Thanks for the :+1: @sgomes, now I also need to make the CI checks pass.

**Review by yuliyan — APPROVED** [review]
@jsnajdr, thank you for working on this!

I've tested the SSR and everything seems to be working fine. I've noticed that there're a few places that are using the translate method bound to the default I18nCalypso instance, but none of them seems to be relevant to the SSR, so it should be OK.

I think it's good to be merged, once you get the CI checks sorted out.

**Comment by jsnajdr** [comment]
So there were `composite-checkout` tests that were never finishing because the cart was forever in a "loading" state and the `await waitFor( /* find desired element on screen */ )` call in the test never returned.

I took me 4 hours to debug this.

Turns out it was caused by `useTranslate` returning a different `translate` function on each call because `useMemo` had wrong dependencies. This `translate` is then used to create a callback that's passed to `CheckoutProvider`, again a prop that changes all the time. And apparently `CheckoutProvider` can't cope with too many rerenders. There is some race condition and effects that break when called too often. Cc @sirbrillig if you want to investigate that further.

Everything is green now, but I'll postpone merging until next morning as this is a rather major patch and I don't want to walk away after deploying.

**Pull request merged**
